### PR TITLE
Ilan/find best reserve

### LINKED ISSUE #11: Determine best reserve for the trade
1. Chose only 1 reserve.
2. If price is similar decide randomly according to `sha3(timestamp)`.
3. Implement similar price mechanism.
a. configurable epsilon - changes smaller than epsilon are negligable
b. go over all reserves and find cheapest price, and store it on array.
c. find all reserves with eplison from cheapest and store in an array
d. pick a reserve at random.

### PR DESCRIPTION
**Pull request by ilanDoron**
please review.
resolved #11

### CODE REVIEW

**Comment by yaronvel** [comment]
why do I see all these commits? didn't we just discuss it?
please make proper branch!!!

**Comment by ilanDoron** [comment]
my mistake. it was on master.
fixed

File: contracts/KyberNetwork.sol [hunk 1]
```diff
@@ -64,15 +64,51 @@ contract KyberNetwork is Withdrawable, KyberConstants {
...
+            //unlist reserves that are more then EPSILON distance from best rate.
+            if (rates[i] != 0 && rates[i] - rates[bestReserve] > EPSILON)
+            {
+                rates[i] = 0;
+                --numRelevantReserves;
+            }
+        }
+
+        //random to choose from remaining reserve.
+        uint random_number = uint(block.blockhash(block.number-1)) % numRelevantReserves;
```
Comment (yaronvel) [inline contracts/KyberNetwork.sol]: bad name

Comment (yaronvel) [inline contracts/KyberNetwork.sol]: what if num relevant reserves is 0?

File: contracts/KyberNetwork.sol [hunk 2]
```diff
@@ -64,15 +64,51 @@ contract KyberNetwork is Withdrawable, KyberConstants {
...
+                bestRate = rates[i];
                 bestReserve = i;
             }
+            ++numRelevantReserves;
+        }
+
+        for ( i = 0; i < numReserves; i++)
+        {
+            //unlist reserves that are more then EPSILON distance from best rate.
+            if (rates[i] != 0 && rates[i] - rates[bestReserve] > EPSILON)
```
Comment (yaronvel) [inline contracts/KyberNetwork.sol]: `{` should be in the first line

Comment (yaronvel) [inline contracts/KyberNetwork.sol]: EPSILON should be configurable and in bps.

File: contracts/KyberNetwork.sol [hunk 3]
```diff
@@ -64,15 +64,51 @@ contract KyberNetwork is Withdrawable, KyberConstants {
...
-        uint rate = 0;
+        uint [] memory rates = new uint[](numReserves);
+        uint numRelevantReserves = 0;
+
         for( uint i = 0 ; i < numReserves ; i++ ) {
-            if( ! (perReserveListedPairs[reserves[i]])[keccak256(source,dest)] ) continue;
+            //list all reserves that have this token.
+            if( ! (perReserveListedPairs[reserves[i]])[keccak256(source,dest)] )
+            {
+                rates[i] = 0;
```
Comment (yaronvel) [inline contracts/KyberNetwork.sol]: everything is init to 0 byte default.

File: contracts/KyberNetwork.sol [hunk 4]
```diff
@@ -64,15 +64,51 @@ contract KyberNetwork is Withdrawable, KyberConstants {
...
 
-            rate = reserves[i].getConversionRate( source, dest, srcQty, block.number );
-            if( rate > bestRate ) {
-                bestRate = rate;
+            rates[i] = reserves[i].getConversionRate( source, dest, srcQty, block.number );
+            if( rates[i] > bestRate ) {
+                bestRate = rates[i];
                 bestReserve = i;
             }
+            ++numRelevantReserves;
```
Comment (yaronvel) [inline contracts/KyberNetwork.sol]: weird way of setting this.

**Pull request merged**